multi/mcxn94x-multi/cpu.c

```diff
@@ -0,0 +1,275 @@
+#include "dev.h"
+
+#define REG_IRQ_OFFSET 4
+/* clang-format off */
+enum {
+	reg_irq = 0, reg_irq_set, reg_irq_clr, mutex = 62
+};
+/* clang-format on */
+
+#define REG_IRQ(CPUID, REG) (*(common.mailbox + (REG_IRQ_OFFSET * (CPUID)) + (REG)))
+
+
+static struct {
+	uint8_t rx_buf[64];
+	lf_fifo_t rx_fifo;
+	handle_t rx_lock;
+	handle_t tx_lock;
+	handle_t irq_lock;
+	handle_t rx_cond;
+
+	volatile uint32_t *mailbox;
+	unsigned int major;
+	uint8_t coreid_self;
+	uint8_t coreid_remote;
+} common;
+
+
+static int mailbox_irqHandler(unsigned int n, void *arg)
+{
+	size_t consumed = 0;
+	uint32_t value = REG_IRQ(common.coreid_self, reg_irq);
+
+	if (value != 0) {
+		REG_IRQ(common.coreid_self, reg_irq_clr) = value;
+
+		size_t rx_len = (value >> 24) & 0x3;
+
+		for (; consumed < rx_len; consumed++) {
+			uint8_t byte = (value >> (8 * consumed)) & 0xff;
+
+			/* TODO: Data is overwritten if rx fifo is full. Consider disabling
+			the interrupt somehow and leaving the value in the register until
+			the application reads from rx fifo. Also, fifo needs to be more
+			generic and at least have function to get its current length. */
+			lf_fifo_push(&common.rx_fifo, byte);
+		}
+	}
+
+	return (consumed > 0) ? 1 : -1;
+}
+
+
+static ssize_t fifoRead(unsigned char *buf, size_t size, unsigned int blocking)
+{
+	size_t bytes_read = 0;
+
+	if (buf == NULL || size == 0) {
+		return -EINVAL;
+	}
+
+	/* Synchronize multiple readers */
+	mutexLock(common.rx_lock);
+
+	if (blocking != 0) {
+		while (bytes_read < size) {
+			mutexLock(common.irq_lock);
+			while (lf_fifo_empty(&common.rx_fifo) != 0) {
+				condWait(common.rx_cond, common.irq_lock, 0);
+			}
+			mutexUnlock(common.irq_lock);
+
+			while (bytes_read < size) {
+				if (lf_fifo_pop(&common.rx_fifo, &buf[bytes_read]) != 0) {
+					bytes_read += 1;
+				}
+				else {
+					break;
+				}
+			}
+		}
+	}
+	else {
+		if (lf_fifo_empty(&common.rx_fifo) != 0) {
+			mutexUnlock(common.rx_lock);
+			return -EWOULDBLOCK;
+		}
+
+		while (bytes_read < size) {
+			if (lf_fifo_pop(&common.rx_fifo, &buf[bytes_read]) != 0) {
+				bytes_read += 1;
+			}
+			else {
+				break;
+			}
+		}
+	}
+
+	mutexUnlock(common.rx_lock);
+
+	return (ssize_t)bytes_read;
+}
+
+
+static ssize_t fifoWrite(const unsigned char *buf, size_t len, unsigned int blocking)
+{
+	size_t to_write;
+	size_t bytes_written = 0;
+
+	if (buf == NULL || len == 0) {
+		return -EINVAL;
+	}
+
+	/* Synchronize multiple writers */
+	mutexLock(common.tx_lock);
+
+	while (bytes_written < len) {
+		if (blocking == 0 && REG_IRQ(common.coreid_remote, reg_irq) != 0u) {
+			mutexUnlock(common.tx_lock);
+			return (bytes_written > 0) ? bytes_written : -EAGAIN;
+		}
+
+		while (blocking != 0 && REG_IRQ(common.coreid_remote, reg_irq) != 0u) {
+			; /* Wait until the other CPU reads IRQx register */
+		}
+
+		/* Message structure: MSB[ 6b reserved | 2b len(data) | 24b data ]LSB */
+		uint32_t value = 0;
+		to_write = min(len - (size_t)bytes_written, 3);
```

Comment:
BTW I would like to suggest consistent use of camelCase ;) 
```suggestion
		size_t to_write = min(len - (size_t)bytes_written, 3);
```

Reply:
Done, camelCased everything